DSP: nsim: introduce nsim_em11d board with DSP feature

SiyuanCheng-CN opened the pull request:
This pull request is to introduce nsim_em11d board with DSP feature, plus a sample test.

Add nsim_em11d board and its DSP and AGU registers context switch code.
Add DSP thread option, dis/enable context switch function for DSP and AGU registers.
Add a sample test, which is currently specific to ARC. 

AGU registers work with XY memory, this PR is dependent on #48405

evgeniy-paltsev commented on include/zephyr/arch/arc/v2/linker.ld:
```diff
@@ -230,6 +230,20 @@ SECTIONS {
 	_estack = .;
 #endif /* __MWDT_LINKER_CMD__ */
 
+	SECTION_PROLOGUE(xdata,,)
+	{
+		__xdata_start = .;
+		*(.Xdata*)
+		__xdata_end = .;
+	} GROUP_DATA_LINK_IN(XCCM, RAMABLE_REGION)
+
+	SECTION_PROLOGUE(ydata,,)
+	{
+		__ydata_start = .;
+		*(.Ydata*)
+		__ydata_end = .;
+	} GROUP_DATA_LINK_IN(YCCM, RAMABLE_REGION)
+
```
I'm wondering if it's OK that we don't do any ifdef guard against XY presence / DSP presence. Won't it break platforms without XY?

evgeniy-paltsev reviewed (changes requested):
Hi @SiyuanCheng-CN,

lgtm, though there are several places for improvement.

stephanosio commented on arch/Kconfig:
```diff
@@ -812,6 +817,24 @@ config FPU_SHARING
 
 endmenu
 
+menu "Digital Signal Processing Options"
+
+config DSP
+	bool "This option enables digital signal processing (DSP)"
+	depends on CPU_HAS_DSP
+	help
+	  This option enables DSP and DSP instructions.
+
+config DSP_SHARING
+	bool "DSP register sharing"
+	depends on DSP && MULTITHREADING
+	help
+	  This option enables preservation of the hardware DSP registers
+	  across context switches to allow multiple threads to perform concurrent
+	  DSP operations.
+
+endmenu
```
Why do we need `DSP` and `DSP_SHARING`?

So far, in every arch we support, the "DSP"/SIMD registers are considered an extension to the FPU context and we have been doing ok with that model.

Why does ARC need to be special? Why can't it just use `FPU` and `FPU_SHARING` with some ARC-specific Kconfigs for configuring what FPU/SIMD/DSP registers ought to be saved?

Also, having both `FPU` and `DSP` can be very confusing on many existing architectures; for example, on ARM, is NEON considered "FPU" or "DSP" (NEON registers extend the FPU registers on ARM and "NEON context" is considered part of "FPU context").

ruuddw replied:
The DSP extensions introduced here are about (efficient HW support for) [fixed point math](https://en.wikipedia.org/wiki/Fixed-point_arithmetic), not floating point. 'hiding' them under FPU doesn't sound right to me.

stephanosio replied:
are the "DSP" registers completely separate from the "FPU registers"?

ARM has a "DSP extension" which can be used for both fixed point math and floating point math, but it uses both arch and FPU registers.

If this ARC "DSP" feature has its own completely independent register file not used by the FPU and its instructions at all, and it only supports fixed point math, then that would be something we have not seen before (in fact, ARC might be the only architecture that does that); but, in that case, these `DSP` and `DSP_SHARING` configs should be under ARC.

ruuddw replied:
ARC is highly configurable, yes it is possible to have the (fixed point) DSP extensions and no floating point. These are independent features. Let us have a closer look at how it's handled in other architectures and see if it makes sense to further align or make this an ARC only feature.

stephanosio replied:
The main question is: "does this DSP feature/unit have its own dedicated register file?"

For ARM, the "DSP extension" uses the arch GP register file for fixed-point math (I mean they are really just integer operations) and VFP register file for floating-point math.

That is, in fact, a quite natural decision to make while you are designing an ISA. So is ARC different in that aspect?

ruuddw replied:
ARC has many more DSP features. There are extra MAC registers for the fixed point, as well as some control registers that determine rounding behavior (saturate/round/trunc). Furthermore, there is a nice feature called AGU: address generation unit. That adds extra registers that are automatically load/stored after accessing (with configurable address increments). That eliminates the need for separate load/store instructions. The inner loop of DSP functions becomes 'mac; mac; ... mac' instead of (load; load; mac; store;)*
That's why technically it could be put under FPU, but conceptually that doesn't make sense.

stephanosio replied:
Ok, in that case, I really think this needs to be part of `arch/arc` and not a kernel-wide feature. In general, both `DSP` and `DSP_SHARING` are quite vague and confusing when it comes to other architectures and what it is supposed to do and/or achieve is not very clear.

ruuddw replied:
The DSP and DSP_SHARING could be introduced locally in arch/arc, how about the status bits in include/zephyr/kernel.h for the DSP sharing? Defining these ARC-local would split the definition, I'd like to avoid that.

stephanosio replied:
> how about the status bits in include/zephyr/kernel.h for the DSP sharing? Defining these ARC-local would split the definition, I'd like to avoid that.

Sure, they will need to live in `kernel.h` inside ifdef ARC. wW have `K_SSE_REGS` for x86 there after all.

stephanosio reviewed (dismissed):
This PR is doing much more than just "introducing nsim_em11d board with DSP features" -- i.e. it is trying to introduce a new "DSP" infrastructure into the kernel.

Generally, I do not agree with the direction this PR is taking adding a separate "DSP" infrastructure.

SiyuanCheng-CN commented on arch/arc/Kconfig:
```diff
@@ -337,6 +337,50 @@ config ARC_NORMAL_FIRMWARE
+	help
+	  Save and restore large AGU registers, including 12 address pointers regs,
+	  8 address offset regs and 24 modifiers regs.
+
```
fixed

evgeniy-paltsev commented on arch/arc/include/swap_macros.h:
```diff
@@ -86,9 +89,14 @@
 #endif
-
+fpu_skip_save :
+#ifdef CONFIG_ARC_DSP_SHARING
```
We do check against CONFIG_ARC_DSP_SHARING in the `_save_dsp_regs` macro itself. So let's keep the check only in the macro and drop it from here - so we'd get a bit cleaner code.

SiyuanCheng-CN replied:
done

evgeniy-paltsev commented on arch/arc/include/swap_macros.h:
```diff
@@ -86,9 +89,14 @@
+	_save_dsp_regs
+#endif
+#ifdef CONFIG_AGU_SHARING
+	_save_agu_regs
```
AGU is a part of DSP - so let's call `_save_agu_regs` from `_save_dsp_regs` macro in `arch/arc/core/dsp/swap_dsp_macros.h` directly.

evgeniy-paltsev commented on arch/arc/include/swap_macros.h:
```diff
@@ -125,9 +132,14 @@
-2 :
 #endif
-
+fpu_skip_load :
+#ifdef CONFIG_ARC_DSP_SHARING
+	_load_dsp_regs
+#endif
+#ifdef CONFIG_AGU_SHARING
+	_load_agu_regs
+#endif
```
Ditto.

evgeniy-paltsev commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -23,6 +23,14 @@ else()
+  zephyr_compile_options_ifdef(CONFIG_SOC_NSIM_EM11D -arcv2em -core3 -Xdiv_rem=radix2
+			       -Xmpy_option=mpyd -Xbitscan -Xswap -Xbarrel_shifter
+			       -Xshift_assist -Xdsp2 -Xdsp_complex -Xxy -Xagu_large
+             -Xdsp_ctrl=postshift,noguard,convergent -Hfxapi
```
Few tabs are missing :)

evgeniy-paltsev commented on arch/arc/Kconfig:
```diff
@@ -337,6 +337,23 @@ config ARC_NORMAL_FIRMWARE
 	  resources of the ARC processors, and, therefore, it shall avoid
 	  accessing them.
 
+config ARC_HAS_DSP
+	bool
+	help
+	  This option is enabled when the ARC CPU has hardware DSP unit.
+
+menu "ARC DSP Options"
+depends on ARC_HAS_DSP
+
+config ARC_DSP
+	bool "This option enables digital signal processing (DSP)"
+	help
+	  This option enables DSP and DSP instructions.
```
Let's move it also to `arch/arc/core/dsp/Kconfig` - so we only have `source "arch/arc/core/dsp/Kconfig"` here.

evgeniy-paltsev commented on soc/arc/snps_nsim/Kconfig.defconfig.em11d:
```diff
@@ -0,0 +1,65 @@
+config HARVARD
+	default y
+
+config ARC_FIRQ
+	default y
```
Not related to DSP directly - I'm currently reworking firq functionality in Zephyr, so let's not add new simulation platforms with firq without necessity. So, let's have `config ARC_FIRQ -> default n`, `config RGF_NUM_BANKS -> default 1`.

And we need to do corresponding change in `nsim_em11d.props` and `mdb_em11d.args`

Thanks!

evgeniy-paltsev commented on tests/arch/arc/arc_dsp_sharing/src/test_common.h:
```diff
@@ -0,0 +1,21 @@
+#if defined(CONFIG_ISA_ARCV2)
+#define THREAD_DSP_FLAGS		(K_DSP_REGS | K_AGU_REGS)
+#else
+#define THREAD_DSP_FLAGS		(K_DSP_REGS)
```
Is it worth to have this 'else' branch? We support this test for ARCv2 only, as I can see from `tests/arch/arc/arc_dsp_sharing/testcase.yaml`

evgeniy-paltsev commented on tests/arch/arc/arc_dsp_sharing/src/load_store.c:
```diff
@@ -0,0 +1,239 @@
+
+		for (i = 0; i < SIZEOF_DSP_REGISTER_SET; i++) {
+			if (store_ptr[i] != init_byte) {
+				TC_ERROR("Found 0x%x instead of 0x%x @ "
+					 "offset 0x%x\n",
+					 store_ptr[i],
+					 init_byte, i);
+				TC_ERROR("Discrepancy found during "
+					 "iteration %d\n",
+					 load_store_low_count);
```
Zephyr now allows 100 character per line - so we can avoid some some string splitting:
```
				TC_ERROR("Found 0x%x instead of 0x%x @offset 0x%x\n",
					 store_ptr[i], init_byte, i);
				TC_ERROR("Discrepancy found during iteration %d\n",
					 load_store_low_count);
```

evgeniy-paltsev reviewed (changes requested):
Hi @SiyuanCheng-CN,

thanks for changes! Still we have some places for improvement.

ruuddw commented on arch/arc/core/dsp/dsp_offsets.c:
```diff
@@ -0,0 +1,72 @@
+ */
+#ifdef CONFIG_ARC_DSP_SHARING
+GEN_OFFSET_SYM(_callee_saved_stack_t, dsp_ctrl);
+GEN_OFFSET_SYM(_callee_saved_stack_t, acc0_lo);
```
The acc0_lo and hi registers are aliases for r58 and r59, which are already saved/restored in Zephyr when CONFIG_ARC_HAS_ACCL_REGS is defined, or CONFIG_64BIT. I think we need some more clever CONFIG and macro definitions to avoid saving these registers multiple times.
With splitting off the DSP stuff in a different file, it is even harder to spot such duplication.

evgeniy-paltsev replied:
Probably it's enough to select CONFIG_ARC_HAS_ACCL_REGS if CONFIG_ARC_DSP_SHARING is set.

stephanosio commented on arch/arc/core/dsp/Kconfig:
```diff
@@ -0,0 +1,71 @@
+	  Processors with XY memory and AGU registers can configure this
+	  option to accelerate DSP instrctions.
+
+config AGU_SHARING
```
```suggestion
config ARC_AGU_SHARING
```

SiyuanCheng-CN replied:
fixed, thanks for reviewing

stephanosio commented on arch/arc/core/dsp/Kconfig:
```diff
@@ -0,0 +1,71 @@
+depends on ARC_HAS_DSP
+
+config ARC_DSP
+	bool "This option enables digital signal processing (DSP)"
```
```suggestion
	bool "Digital Signal Processing (DSP)"
```

stephanosio commented on arch/arc/include/kernel_arch_func.h:
```diff
@@ -82,7 +82,6 @@ extern FUNC_NORETURN void z_arc_switch_to_main_no_multithreading(
 
 #endif /* !CONFIG_MULTITHREADING */
 
-
```
spurious change

stephanosio reviewed (commented):
Looks ok overall. Just some minor comments.

Commit:
DSP: add DSP support for ARC processor

add DSP reg in context switch
add AGU reg in context switch to support XY mem
add thread option and API to dis/enable DSP switch

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

Commit:
arc: add nsim_em11d target

add nsim_em11d target specific for DSP feature

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

Commit:
DSP: add dsp unit test

add dsp context switch test
add complex multiplication test for ARC processor

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

evgeniy-paltsev commented on include/zephyr/kernel.h:
```diff
@@ -213,6 +213,38 @@ extern void k_thread_foreach_unlocked(
 
+#ifdef CONFIG_ARC
+/* ARC processor Bitmask definitions for threads user options */
+#define K_FP_IDX 1
```
Let's move it from ARC options to K_FP_REGS definition itself. So we'll get:

```
#if defined(CONFIG_FPU_SHARING)
/**
 * @brief FPU registers are managed by context switch
 *
 * @details
 * This option indicates that the thread uses the CPU's floating point
 * registers. This instructs the kernel to take additional steps to save
 * and restore the contents of these registers when scheduling the thread.
 * No effect if @kconfig{CONFIG_FPU_SHARING} is not enabled.
 */
#define K_FP_IDX 1
#define K_FP_REGS (BIT(K_FP_IDX))
#endif
```

in `include/zephyr/kernel.h`

evgeniy-paltsev commented on arch/arc/core/dsp/swap_dsp_macros.h:
```diff
@@ -0,0 +1,270 @@
+	ld_s r13, [r2, ___thread_base_t_user_options_OFFSET]
+	bbit0 r13, K_DSP_IDX, dsp_skip_save
+	lr r13, [_ARC_V2_DSP_CTRL]
+	st_s r13, [sp, ___callee_saved_stack_t_dsp_ctrl_OFFSET]
+	lr r13, [_ARC_V2_ACC0_LO]
```
From my understanding It's not entirely correct.

The ACCL/ACCH regs are used not only by 'pure' DSP instructions - they are also touched by multiply (for example) instruction which can be generated in other kernel / user code. So it's not correct to save / restore ACCL/ACCH only if `K_DSP` bit is set for thread.

Probably the best solution is to drop  ACCL/ACCH management from DSP code and select CONFIG_ARC_HAS_ACCL_REGS if CONFIG_ARC_DSP_SHARING is set.

Note that we still need to save / restore ACC0_GLO & ACC0_GHI here.

SiyuanCheng-CN replied:
fixed

evgeniy-paltsev commented on arch/arc/core/dsp/Kconfig:
```diff
@@ -0,0 +1,73 @@
+config ARC_DSP_BFLY_SHARING
+	bool "ARC complex DSP operation"
+	depends on ARC_DSP && CPU_ARCEM
+	default n
```
`n` is already default, we can drop `default n`

evgeniy-paltsev commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -23,6 +23,14 @@ else()
 			       -Xfpus_div -Xfpu_mac -Xfpuda -Xfpus_mpy_slow
 			       -Xfpus_div_slow -Xbitstream -Xtimer0 -Xtimer1)
 
+  zephyr_compile_options_ifdef(CONFIG_SOC_NSIM_EM11D -arcv2em -core3 -Xdiv_rem=radix2
+			       -Xmpy_option=mpyd -Xbitscan -Xswap -Xbarrel_shifter
+			       -Xshift_assist -Xdsp2 -Xdsp_complex -Xxy -Xagu_large
+			       -Xdsp_ctrl=postshift,noguard,convergent -Hfxapi
+			       -Xdsp_divsqrt=radix2 -Xdsp_itu -Xdsp_accshift=full
+			       -Xfpus_div -Xfpu_mac -Xfpuda -Xfpus_mpy_slow
+			       -Xfpus_div_slow -Xbitstream -Xtimer0 -Xtimer1)
```
Another issue I see here is that we pass DSP options to generic C flags - so some DSP instructions may be generated while building kernel / application so we may execute DSP instructions in threads which are not marked with 'K_DSP_REGS' flag.

We need somehow distinguish flags for kernel / non-dsp code and dsp code.
Probably we can drop DSP flags from here and just provide DSP flag list in some Cmake variable, so application code can easily add it to the base C flags when building DSP code part.

ruuddw commented on include/zephyr/arch/arc/v2/dsp/arc_dsp.h:
```diff
@@ -0,0 +1,41 @@
+ * The function is used to enable the preservation of dsp
+ * context information for a particular thread.
+ * This API depends on each architecture implementation. If the architecture
+ * does not support enabling, this API will always be failed.
```
I don't see any error handling or so in the implementation. How will it fail?
Furthermore, the implementation doesn't check if only the DSP sharing related fields are modified. So either the description or the implementation should change.

ruuddw commented on arch/arc/core/dsp/dsp_offsets.c:
```diff
@@ -0,0 +1,72 @@
+ */
+#ifdef CONFIG_ARC_DSP_SHARING
+GEN_OFFSET_SYM(_callee_saved_stack_t, dsp_ctrl);
+GEN_OFFSET_SYM(_callee_saved_stack_t, acc0_lo);
```
Are acc0_lo/hi required? These are mirrors of the accl core registers.

ruuddw commented on arch/arc/include/kernel_arch_data.h:
```diff
@@ -158,7 +158,69 @@ struct _callee_saved_stack {
+	uintptr_t dsp_bfly0;
+#endif
+	uintptr_t acc0_ghi;
+	uintptr_t acc0_hi;
```
Are acc0_lo/hi required? These are mirrors of the accl core registers.

evgeniy-paltsev commented on include/zephyr/kernel.h:
```diff
@@ -213,6 +214,37 @@ extern void k_thread_foreach_unlocked(
+ * restore the contents of these registers when scheduling the thread.
+ * No effect if @kconfig{CONFIG_ARC_DSP_SHARING} is not enabled.
+ */
+#define K_DSP_IDX 6
+#define K_DSP_REGS (BIT(K_DSP_IDX))
+#endif
```
I'd propose to define ARC-specific bits with arc prefix, K_DSP_REGS -> K_ARC_DSP_REGS.
(same for indexes and for K_AGU_REGS)

evgeniy-paltsev commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -23,6 +23,17 @@ else()
 
+  zephyr_compile_options_ifdef(CONFIG_SOC_NSIM_EM11D -arcv2em -core3 -Xdiv_rem=radix2
+			       -Xmpy_option=mpyd -Xbitscan -Xswap -Xbarrel_shifter
+			       -Xshift_assist -Xdsp2 -Xdsp_complex
+			       -Xdsp_divsqrt=radix2 -Xdsp_itu -Xdsp_accshift=full
```
Hmm, do we sure that instructions generated after we pass `-Xdsp2` `-Xdsp_complex` `-Xdsp_itu` `-Xdsp_accshift=full` there will be no instructions generated which can change any value except 64-bit accumulator (which we do save/restore every time).

SiyuanCheng-CN replied:
Also move `-Xmpy_option=mpyd` to dsp options, as it need `-Xdsp2`

evgeniy-paltsev commented on tests/arch/arc/arc_dsp_sharing/src/dsp_regs_arc.h:
```diff
@@ -0,0 +1,99 @@
+ * @brief ARC specific dsp register macros
+ */
+
+/*
+ * Copyright (c) 2022, Synopsys.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Ditto

evgeniy-paltsev commented on arch/arc/core/thread.c:
```diff
@@ -288,3 +292,29 @@ FUNC_NORETURN void z_arc_switch_to_main_no_multithreading(k_thread_entry_t main_
+void arc_dsp_disable(struct k_thread *thread, unsigned int options)
+{
+	/* Ensure a preemptive context switch does not occur */
+
```
Let's drop extra line between code and comment

evgeniy-paltsev commented on arch/arc/core/thread.c:
```diff
@@ -288,3 +292,29 @@ FUNC_NORETURN void z_arc_switch_to_main_no_multithreading(k_thread_entry_t main_
+void arc_dsp_enable(struct k_thread *thread, unsigned int options)
+{
+	/* Ensure a preemptive context switch does not occur */
+
```
ditto

evgeniy-paltsev commented on tests/arch/arc/arc_dsp_sharing/src/dsp_regs_arc.h:
```diff
@@ -0,0 +1,99 @@
+{
+	uint32_t temp = 0;
+
+	__asm__ volatile (
+		"ld %1, [%0, 0];\n\t"
+		"sr %1, [%2];\n\t"
+		"ld %1, [%0, 4];\n\t"
+		"sr %1, [%3];\n\t"
+		"ld %1, [%0, 8];\n\t"
+		"sr %1, [%4];\n\t"
+		: : "r" (regs), "r" (temp),
+		"i"(_ARC_V2_DSP_BFLY0), "i"(_ARC_V2_AGU_AP0),
+		"i"(_ARC_V2_AGU_OS0)
+		: "memory"
+		);
```
Hmm, I'm wondering if we actually need inline ASM in _load_all_dsp_registers / _store_all_dsp_registers or we can simply replace it ith aux_read / aux write?

SiyuanCheng-CN replied:
Instructions related to aux can be replaced with aux_read / aux write, but we still need `st` and `ld` to write and read from stack.

evgeniy-paltsev replied:
> Instructions related to aux can be replaced with aux_read / aux write, but we still need st and ld to write and read from stack.

As I can see we are not reading / writing from stack - we are reading from dsp_register_set structure. So we can just read  / write from/to member of this structure. I don't see any reason to do it in ASM.

Please correct me if I'm wrong.

evgeniy-paltsev commented on boards/arc/nsim/nsim_em11d.yaml:
```diff
@@ -0,0 +1,11 @@
+simulation: nsim
+arch: arc
+toolchain:
+  - arcmwdt
```
I'd propose to not limit the platform to `arcmwdt`, instead let's limit the `test_calculation` dsp testcase to `arcmwdt`.

The rest of the thing can be build/run with GNU toolchain. 
Of course such variant (`nsim_em11d` + GNU toolchain) is currently mostly for testing and not for any real use cases.

To implement this we probably need to add EM GNU toolchain options and keep `EM11D_DSP_OPTIONS` empty for GNU toolchain.

SiyuanCheng-CN replied:
For this test, nsim_em11d + GNU toolchain doesn't need EM GNU toolchain options, as there is no DSP operation in `test_load_store`. Now only enabled `EM11D_DSP_OPTIONS` for mwdt toolchain.

evgeniy-paltsev commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -23,6 +23,17 @@ else()
+			       -Xfpus_div_slow -Xbitstream -Xtimer0 -Xtimer1)
+
+  if(CONFIG_SOC_NSIM_EM11D)
+    set(ENV{EM11D_DSP_OPTIONS} "-Xxy;-Xagu_large;-Xdsp_ctrl=postshift,noguard,convergent;-Hfxapi")
```
Let's use more generic naming so we can avoid extra `if`'s in test's cmake when we add more ARC DSP-enabled configurations. Like `EM11D_DSP_OPTIONS` -> `Z_ARC_DSP_OPTIONS`.

evgeniy-paltsev commented on tests/arch/arc/arc_dsp_sharing/testcase.yaml:
```diff
@@ -0,0 +1,8 @@
+    slow: true
+    tags: dsp arc
+    timeout: 600
+    min_ram: 16
```
This looks like a copy-paste from `tests/kernel/fpu_sharing/generic/testcase.yaml`. :)
If it's so and we haven't do some memory estimations for this test - let's simply drop it.

evgeniy-paltsev commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -23,6 +23,16 @@ else()
+			       -Xfpus_div_slow -Xbitstream -Xtimer0 -Xtimer1)
+
+  if(CONFIG_SOC_NSIM_EM11D)
+    set(ENV{Z_ARC_DSP_OPTIONS} "-Xxy;-Xagu_large;-Hfxapi;-Xdsp2;-Xdsp_complex;-Xdsp_accshift=full;\
+-Xdsp_divsqrt=radix2;-Xdsp_itu;-Xdsp_ctrl=postshift,noguard,convergent;-Xmpy_option=mpyd")
```
I'd rather avoid environment usage for internal variables. We can use properties instead. I.e:

```suggestion
    set_property(GLOBAL PROPERTY z_arc_dsp_options "-Xxy;-Xagu_large;-Hfxapi;-Xdsp2;-Xdsp_complex;-Xdsp_accshift=full;\
-Xdsp_divsqrt=radix2;-Xdsp_itu;-Xdsp_ctrl=postshift,noguard,convergent;-Xmpy_option=mpyd")
```

evgeniy-paltsev commented on tests/arch/arc/arc_dsp_sharing/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+      ${ARCMWDT_TOOLCHAIN_PATH}/MetaWare/arc/lib/src/fx/include/
+    )
+target_sources(app PRIVATE src/calculation_arc.c)
+target_compile_options(app PRIVATE $ENV{Z_ARC_DSP_OPTIONS})
```
```suggestion
get_property(Z_ARC_DSP_OPTIONS GLOBAL PROPERTY z_arc_dsp_options)
target_compile_options(app PRIVATE ${Z_ARC_DSP_OPTIONS})
```

evgeniy-paltsev commented on arch/arc/core/dsp/swap_dsp_macros.h:
```diff
@@ -0,0 +1,262 @@
+	lr r13, [_ARC_V2_AGU_MOD20]
+	st r13, [sp, ___callee_saved_stack_t_agu_mod20_OFFSET]
+	lr r13, [_ARC_V2_AGU_MOD21]
+	_st32_huge_offset r13, sp, ___callee_saved_stack_t_agu_mod21_OFFSET, r1
```
I'm really curious about this - for both st and ld we have same offset size here (s9 value). So how we do 
`ld r13, [sp, ___callee_saved_stack_t_agu_mod21_OFFSET]` 
but can't do 
`st r13, [sp, ___callee_saved_stack_t_agu_mod21_OFFSET]`

?

ruuddw reviewed (dismissed):
Thanks, looks ready to be merged to me now.

stephanosio commented on arch/arc/core/dsp/Kconfig:
```diff
@@ -0,0 +1,72 @@
+config ARC_AGU_MEDIUM
+	bool "ARC AGU medium size register"
+	depends on ARC_AGU_SHARING
+	default n
```
default n is redundant.

stephanosio commented on arch/arc/core/dsp/Kconfig:
```diff
@@ -0,0 +1,72 @@
+config ARC_AGU_LARGE
+	bool "ARC AGU large size register"
+	depends on ARC_AGU_SHARING
+	default n
```
Ditto

stephanosio commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -23,6 +23,16 @@ else()
+			       -Xfpus_div_slow -Xbitstream -Xtimer0 -Xtimer1)
+
+  if(CONFIG_SOC_NSIM_EM11D)
+    set_property(GLOBAL PROPERTY z_arc_dsp_options "-Xxy;-Xagu_large;-Hfxapi;-Xdsp2;-Xdsp_accshift=full;\
```
Why is z_arc_dsp_options necessary?

evgeniy-paltsev reviewed (changes requested):
Please don't do line wraparound where it's not required as it makes code less readable.

evgeniy-paltsev commented:
Hi @SiyuanCheng-CN,

I've tried to build dsp_sharing.test_load_store with GNU toolchain from Zephyr SDK 0.15.0 with following command:

```
west build -b nsim_em11d tests/arch/arc/arc_dsp_sharing -T dsp_sharing.test_load_store
```

and I got multiple warnings on linking stages, like:

```
[130/140] Linking C executable zephyr/zephyr_pre0.elf
/home/paltsev/CodeData/zephyr/zephyr-sdk/zephyr-sdk-0.15.0/arc-zephyr-elf/bin/../lib/gcc/arc-zephyr-elf/12.1.0/../../../../arc-zephyr-elf/bin/ld.bfd:zephyr/linker_zephyr_pre0.cmd:250: warning: memory region `XCCM' not declared
/home/paltsev/CodeData/zephyr/zephyr-sdk/zephyr-sdk-0.15.0/arc-zephyr-elf/bin/../lib/gcc/arc-zephyr-elf/12.1.0/../../../../arc-zephyr-elf/bin/ld.bfd:zephyr/linker_zephyr_pre0.cmd:257: warning: memory region `YCCM' not declared

```

I'm wondering if it behaves same way for you?

SiyuanCheng-CN commented:
> Hi @SiyuanCheng-CN,
> 
> I've tried to build dsp_sharing.test_load_store with GNU toolchain from Zephyr SDK 0.15.0 with following command:
> 
> ```
> west build -b nsim_em11d tests/arch/arc/arc_dsp_sharing -T dsp_sharing.test_load_store
> ```
> 
> and I got multiple warnings on linking stages, like:
> 
> ```
> [130/140] Linking C executable zephyr/zephyr_pre0.elf
> /home/paltsev/CodeData/zephyr/zephyr-sdk/zephyr-sdk-0.15.0/arc-zephyr-elf/bin/../lib/gcc/arc-zephyr-elf/12.1.0/../../../../arc-zephyr-elf/bin/ld.bfd:zephyr/linker_zephyr_pre0.cmd:250: warning: memory region `XCCM' not declared
> /home/paltsev/CodeData/zephyr/zephyr-sdk/zephyr-sdk-0.15.0/arc-zephyr-elf/bin/../lib/gcc/arc-zephyr-elf/12.1.0/../../../../arc-zephyr-elf/bin/ld.bfd:zephyr/linker_zephyr_pre0.cmd:257: warning: memory region `YCCM' not declared
> ```
> 
> I'm wondering if it behaves same way for you?

Did you add XY mem PR, it's now merged into upstream.

evgeniy-paltsev commented:
> Did you add XY mem PR, it's now merged into upstream.

Nope, for some reason I've thought it was also embedded in this PR, but it's not :)
I'll do rebase locally and re-check.

evgeniy-paltsev commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -23,6 +23,16 @@ else()
+			       -Xfpus_div_slow -Xbitstream -Xtimer0 -Xtimer1)
+
+  if(CONFIG_SOC_NSIM_EM11D)
+    set_property(GLOBAL PROPERTY z_arc_dsp_options "-Xxy;-Xagu_large;-Hfxapi;-Xdsp2;-Xdsp_accshift=full;\
+-Xdsp_divsqrt=radix2;-Xdsp_complex;-Xdsp_itu;-Xdsp_ctrl=postshift,noguard,convergent;-Xmpy_option=mpyd")
```
Any reason we specify compilers options here in such specific manner?

I've tried to specify it normally - and it seems to work fine.

```suggestion
    set_property(GLOBAL PROPERTY z_arc_dsp_options -Xxy -Xagu_large -Hfxapi -Xdsp2
                 -Xdsp_accshift=full -Xdsp_divsqrt=radix2 -Xdsp_complex -Xdsp_itu
                 -Xdsp_ctrl=postshift,noguard,convergent -Xmpy_option=mpyd)
```
